consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/pki/PkiQbftExtraDataCodec.java

```diff
@@ -37,54 +36,77 @@ public BftExtraData decodeRaw(final Bytes input) {
       throw new IllegalArgumentException("Invalid Bytes supplied - Bft Extra Data required.");
     }
 
+    final BftExtraData bftExtraData = super.decodeRaw(input);
+
+    // Skip through every element in the list and read the last one (CMS message)
     final RLPInput rlpInput = new BytesValueRLPInput(input, false);
     rlpInput.enterList();
-
-    // Consume all the ExtraData input from the envelope list, and decode it using the QBFT decoder
-    final Bytes extraDataListAsBytes = rlpInput.currentListAsBytes();
-    final BftExtraData bftExtraData = super.decodeRaw(extraDataListAsBytes);
-
-    final Optional<Bytes> cms;
-    if (rlpInput.nextIsList() && rlpInput.nextSize() == 0) {
-      cms = Optional.empty();
+    rlpInput.skipNext();
+    rlpInput.skipNext();
+    rlpInput.skipNext();
+    rlpInput.skipNext();
+    rlpInput.skipNext();
```

Comment:
Forgive my naive understanding with very limited context...I'm wondering how robust this is to potential future changes...is it possible/likely that this list will change and make this brittle?

Reply:
You are 100%. This implementation is very dependent on the RLP format and the `QbftExtraDataCodec` implementation. However, this is the case for most RLP structures, as they are very rigid (e.g. order dependent, etc.).

The best we can do is to prevent this breaking in the future is to add tests around the decoding/encoding of the expected input to ensure we catch any issues (e.g. if someone changes the "standard" RLP structure or change some details in the `QbftExtraDataCodec` implementation, we fail a unit test to flag that we need to update this class).

Another alternative that we have explored (and might bring back to solve the duplication in the encode() method) is to use an "envelope" list. This won't solve the coupling between this class and the QbftExtraData codec but it will definitely make the code easier to read (it will adhere better to the RLP library API).

Reply:
To add a bit more context. We don't use a key/value structure on our RLP encoding, it is all positional. That also makes things a bit more complicated when parsing and extracting specific values from the RLP byte array :)